Asset Management and Bootstrap

DianeRt opened the pull request:
@gjaldon Done with issues #7 and #8. I had trouble getting the stylesheets and js to work 😅 Turns out I needed a public folder for the bootstrap files and had to correct the sprockets' environment.append_path 😌 
- Added assets folder and sprockets core files
- Added codes in application_controller.rb to make use of sprockets
- Added public folder in root folder to contain bootstrap files
- Edited layout.erb to link make use of bootstrap and sprockets

DianeRt committed:
Added Rakefile and sequel migration

DianeRt committed:
Deleted create_table.rb

DianeRt committed:
Edited Rakefile

DianeRt committed:
Moved file to db/migrations

DianeRt committed:
Added a timestamped Sequel migration generator

DianeRt committed:
Deleted original migration file

DianeRt committed:
Created new migration files for creating users table and adding columns

DianeRt committed:
Renamed and edited crud.rb to application_controller.rb

DianeRt committed:
Created config.ru to run UsersController

DianeRt committed:
Created users_controller.rb for '/users' routes

DianeRt committed:
Renamed crud to application_controller, moved all erb files to app/views and renamed users/index.erb to admin.erb

DianeRt committed:
Added rack gem to Gemfile

DianeRt committed:
Added minitest and nokogiri to Gemfile

DianeRt committed:
Added mini:test to Rakefile

DianeRt committed:
Added configure :test and configure :development

DianeRt committed:
Changed back erb:admin to erb:index

DianeRt committed:
Creates minitest spec files

DianeRt committed:
Renamed index.erb to home.erb and admin.erb to index.erb

gjaldon committed:
add rake and rack-test to test and development gems

DianeRt committed:
Removed DB drop/create/insert from spec_helper.rb

DianeRt committed:
Added test environment option to db:migrate and added db:migrate task to mini:test

DianeRt committed:
Added error 403 and 500. Also disabled show_exceptions.

DianeRt committed:
Changed last_response.body to use .status instead. Added after block to truncate table.

DianeRt committed:
Added sprockets gem to Gemfile

DianeRt committed:
Added sprocket codes to application_controller and layout.erb

DianeRt committed:
Added sprocket core files

DianeRt committed:
Added uglifier and sass gem to Gemfile

DianeRt committed:
Added link to bootstrap stylesheet and js

DianeRt committed:
Corrected sprockets environment.append_path and added set :public_folder

DianeRt committed:
Added public folder with bootstrap files

gjaldon commented:
@DianeRt instead of using sprocket for asset management, let's use a Node build tool instead like [Gulp](http://gulpjs.com). We want similar behavior with sprockets where 1 `app.css` and 1 `app.js` file are produced after building. Some other things we'd like:
- SCSS compilation
- es6 javascript (take a look at Babel for Gulp) > this is so we can use es6 features when writing our javascript
- minification > so that output js and css filesize is smaller. for faster downloads in production
- source maps > even if files are minified, we can view the source version in our browser dev tools

The great thing about using Gulp or some other Node build tool is that we can install our frontend dependencies using NPM. That way, we manage our frontend deps with NPM instead of copying assets into the public folder and commit them into version control.

You will need to learn how to use NPM and Gulp for this task. NPM works just like Bundler/Rubygems.

DianeRt committed:
Deleted bootstrap's public folder and sprockets' assets folders

DianeRt committed:
Removed sprockets and related gems from Gemfile

DianeRt committed:
Removed codes related to sprockets

DianeRt committed:
Removed links from sprockets and added new links to bootstrap, css and js files

DianeRt committed:
Created new view page for errors

DianeRt committed:
Added gulp and related files

DianeRt commented:
I've made changes to the files to use gulp. Gulp watch is amazing! 😲

gjaldon committed:
ignore node_modules and remove from repo

gjaldon commented on app/views/layout.erb:
```diff
@@ -0,0 +1,36 @@
+
+    <!-- Bootstrap -->
+    <!-- using href="css/bootstrap.min.css" causes show.erb to ignore bootstrap -->
+    <link href="/node_modules/bootstrap/dist/css/bootstrap.min.css" rel="stylesheet">
```
Remove this line. This line is not necessary when we build a css file with gulp that already includes bootstrap. This might help you with setting up Gulp to do that:
https://gist.github.com/ericbarnes/ac3ae075c97c1073869c

Note that we only want to use NPM. Do not install Bower.

gjaldon commented:
@DianeRt I added `npm_modules` folder to `.gitignore`. We do not need to commit node_modules to the git repo since we can do `npm install`. This helps to keep our repo small. :)

gjaldon commented on app/controllers/application_controller.rb:
```diff
@@ -0,0 +1,86 @@
+
+configure :development do
+  set :database, 'postgresql://localhost/users'
+  DB = Sequel.connect("postgresql://localhost/users")
```
Change the name of the db to `crud_development` instead of `users`. `users` is usually used to represent a table. For db_names, we usually use the name of the app and the env we use the db in.

gjaldon commented on app/controllers/application_controller.rb:
```diff
@@ -0,0 +1,86 @@
+Bundler.require(:default) 
+
+configure :test do
+ set :database, 'postgresql://localhost/test'
```
Change the name of the db to `crud_test` instead of `test`.

gjaldon commented on Rakefile:
```diff
@@ -0,0 +1,52 @@
+  desc "Run minitest"
+  task :test
+  ENV['RACK_ENV'] = 'test'
+  Rake::Task["db:migrate"].invoke
```
Remove this line. The test will raise an error if a migration is necessary.

gjaldon commented on gulpfile.js:
```diff
@@ -0,0 +1,47 @@
+
+//using babel and yui compressor
+//this will concatenate, minify and add es6 js features to all files in source
+gulp.task('babel', function () {
```
We're currently not using this. Later on, I want us to use https://github.com/babel/babelify when we write some JS for the frontend.

DianeRt committed:
Edited gulpfile to utilize browserify and babelify

DianeRt committed:
Removed invoke command in mini:test task

DianeRt committed:
Changed database names from users to crud_development and test to crud_test

DianeRt committed:
Removed line linking to bootstrap stylesheet and replaced with link to the files created by gulp tasks

DianeRt commented:
@gjaldon  Thanks for adding `node_modules` to `.gitignore` 😄 

Finally worked on pending tasks
- Removed invoke line inside mini:test task in `Rakefile`
- Removed bootstrap link in `layout.erb`
- Renamed database to `crud_test` and `crud_development`
- Added browserify and babelify in `gulpfile.js`

I haven't removed the babel task in gulpfile.js. Should I keep both babel and babelify?
